src/logic/mutable_addon.cc

```diff
@@ -0,0 +1,307 @@
+		}
+	} else {
+		FileRead fr;
+		fr.open(*g_fs, source);
+		const size_t bytes = fr.get_size();
+		std::unique_ptr<char[]> data(new char[bytes]);
+		fr.data_complete(data.get(), bytes);
+
+		FileWrite fw;
+		fw.data(data.get(), bytes);
+		fw.write(*g_fs, dest);
+	}
+}
+
+std::unique_ptr<MutableAddOn> MutableAddOn::create_mutable_addon(const AddOnInfo& a) {
+	switch (a.category) {
+	case AddOnCategory::kWorld:
+		return std::unique_ptr<WorldAddon>(new WorldAddon(a));
+	case AddOnCategory::kTribes:
+		return std::unique_ptr<TribesAddon>(new TribesAddon(a));
+	case AddOnCategory::kScript:
+		return std::unique_ptr<ScriptAddon>(new ScriptAddon(a));
+	case AddOnCategory::kMaps:
+		return std::unique_ptr<MapsAddon>(new MapsAddon(a));
+	case AddOnCategory::kCampaign:
+		return std::unique_ptr<CampaignAddon>(new CampaignAddon(a));
+	case AddOnCategory::kWinCondition:
+		return std::unique_ptr<WinCondAddon>(new WinCondAddon(a));
+	case AddOnCategory::kStartingCondition:
+		return std::unique_ptr<StartingCondAddon>(new StartingCondAddon(a));
+	case AddOnCategory::kTheme:
+		return std::unique_ptr<ThemeAddon>(new ThemeAddon(a));
+	default:
+		return std::unique_ptr<MutableAddOn>(new MutableAddOn(a));
+	}
+}
+
+MutableAddOn::MutableAddOn(const AddOnInfo& a)
+   : internal_name_(a.internal_name),
+     descname_(a.unlocalized_descname),
+     description_(a.unlocalized_description),
+     author_(a.unlocalized_author),
+     version_(AddOns::version_to_string(a.version, false)),
+     category_(a.category),
+     directory_(kAddOnDir + FileSystem::file_separator() + internal_name_) {
+}
+
+std::string MutableAddOn::profile_path() {
+	return directory_ + FileSystem::file_separator() + kAddOnMainFile;
+}
+
+void MutableAddOn::update_info(const std::string& descname,
+                               const std::string& author,
+                               const std::string& description,
+                               const std::string& version) {
+	descname_ = descname;
+	author_ = author;
+	description_ = description;
+	version_ = version;
+}
+
+std::string MutableAddOn::parse_requirements() {
+	// · We need to read the original `addon` file (if it exists) to
+	//   determine the requirements. Then write the file, and we are done.
+	const std::string directory = kAddOnDir + FileSystem::file_separator() + internal_name_;
+	bool dir_exists = g_fs->file_exists(directory);
+	const std::string profile_path = directory + FileSystem::file_separator() + kAddOnMainFile;
+
+	std::string requires;
+	if (dir_exists) {
+		Profile p;
+		p.read(profile_path.c_str());
+		requires = p.get_safe_section("global").get_safe_string("requires");
+	}
+	return requires;
+}
+
+bool MutableAddOn::write_to_disk() {
+	// Step 1: Gather the requirements of all contained maps
+	std::string requires = parse_requirements();
+
+	// Step 2: Create the `addon` file.
+	g_fs->ensure_directory_exists(directory_);
+	// Write profile
```

Comment:
Add a newline and prefix the comment with "Step 3:"
Also, no need to enclose the profile code in {} now.